src/query_plan/operation.rs

```diff
@@ -0,0 +1,426 @@
+                    selection_set: SelectionSet {
+                        ty: normalized_field.selection_set.ty.clone(),
+                        selections,
+                    },
+                };
+                flattened.push(Selection::Field(Node::new(field)));
+            }
+            NormalizedSelection::NormalizedInlineFragment(normalized_fragment) => {
+                let selections = flatten_selections(&normalized_fragment.selection_set.selections);
+                let fragment = InlineFragment {
+                    type_condition: normalized_fragment.type_condition.to_owned(),
+                    directives: normalized_fragment.directives.to_owned(),
+                    selection_set: SelectionSet {
+                        ty: normalized_fragment.selection_set.ty.clone(),
+                        selections,
+                    },
+                };
+                flattened.push(Selection::InlineFragment(Node::new(fragment)));
+            }
+        }
+    }
+    flattened
+}
+
+/// Normalizes selection set within specified operation.
+///
+/// This method applies following normalizations
+/// - expands all fragments within an operation
+/// - merge same selections
+/// - removes all introspection fields from top-level selection
+/// - attempts to remove all unnecessary/redundant inline fragments
+pub fn normalize_operation(
+    operation: &mut Operation,
+    _schema: &Schema,
+    fragments: &IndexMap<Name, Node<Fragment>>,
+) {
+    let mut normalized_selection_set =
+        NormalizedSelectionSet::from_selection_set(&operation.selection_set, fragments);
+    // removes top level introspection
+    normalized_selection_set
+        .selections
+        .retain(|key, _| !key.starts_with("__"));
+
+    // flatten back to vec
+    operation.selection_set = SelectionSet::from(normalized_selection_set);
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::query_plan::operation::normalize_operation;
+    use apollo_compiler::executable::Name;
+    use apollo_compiler::NodeStr;
+
+    #[test]
+    fn expands_named_fragments() {
+        let operation_with_named_fragment = r#"
+query NamedFragmentQuery {
+  foo {
+    id
+    ...Bar
+  }
+}
+
+fragment Bar on Foo {
+  bar
+  baz
+}
+
+type Query {
+  foo: Foo
+}
+
+type Foo {
+  id: ID!
+  bar: String!
+  baz: Int
+}
+"#;
+        let (schema, mut executable_document) =
+            apollo_compiler::parse_mixed(operation_with_named_fragment, "document.graphql");
+
+        if let Some(operation) = executable_document
+            .named_operations
+            .get_mut(&Name::new_unchecked(NodeStr::new("NamedFragmentQuery")))
+        {
+            let operation = operation.make_mut();
+            normalize_operation(operation, &schema, &executable_document.fragments);
+
+            let expected = r#"query NamedFragmentQuery {
+  foo {
+    id
+    bar
+    baz
+  }
+}"#;
+            let actual = format!("{}", operation);
```

Comment:
Equivalent and maybe a bit easier to type:
```suggestion
            let actual = operation.to_string();
```